server/protocols/jmap-rfc-8621/src/main/scala/org/apache/james/jmap/routes/DownloadRoutes.scala

```diff
@@ -0,0 +1,240 @@
+import reactor.core.scheduler.Schedulers
+import reactor.netty.http.server.{HttpServerRequest, HttpServerResponse}
+
+import scala.compat.java8.FunctionConverters._
+import scala.jdk.CollectionConverters._
+import scala.util.{Failure, Success, Try}
+
+object DownloadRoutes {
+  val LOGGER: Logger = LoggerFactory.getLogger(classOf[DownloadRoutes])
+
+  val BUFFER_SIZE: Int = 16 * 1024
+}
+
+trait BlobResolver {
+  def resolve(blobId: BlobId, mailboxSession: MailboxSession): Option[SMono[Blob]]
+}
+
+trait Blob {
+  def blobId: BlobId
+  def contentType: ContentType
+  def size: Try[Size]
+  def content: InputStream
+}
+
+case class BlobNotFoundException(blobId: BlobId) extends RuntimeException
+
+case class MessageBlob(blobId: BlobId, message: MessageResult) extends Blob {
+  override def contentType: ContentType = new ContentType("message/rfc822")
+
+  override def size: Try[Size] = refineV[NonNegative](message.getSize) match {
+    case Left(e) => Failure(new IllegalArgumentException(e))
+    case Right(size) => Success(size)
+  }
+
+  override def content: InputStream = message.getFullContent.getInputStream
+}
+
+case class EmailBodyPartBlob(blobId: BlobId, part: EmailBodyPart) extends Blob {
+  override def size: Try[Size] = Success(part.size)
+
+  override def contentType: ContentType = new ContentType(part.`type`.value)
+
+  override def content: InputStream = {
+    val writer = new DefaultMessageWriter
+    val outputStream = new ByteArrayOutputStream()
+    writer.writeBody(part.entity.getBody, outputStream)
+    new ByteArrayInputStream(outputStream.toByteArray)
+  }
+}
+
+class MessageBlobResolver @Inject()(val messageIdFactory: MessageId.Factory,
+                          val messageIdManager: MessageIdManager) extends BlobResolver {
+  override def resolve(blobId: BlobId, mailboxSession: MailboxSession): Option[SMono[Blob]] = {
+    Try(messageIdFactory.fromString(blobId.value.value)) match {
+      case Failure(_) => None
+      case Success(messageId) => Some(SMono.fromPublisher(
+        messageIdManager.getMessagesReactive(List(messageId).asJava, FetchGroup.FULL_CONTENT, mailboxSession))
+        .map[Blob](MessageBlob(blobId, _))
+        .switchIfEmpty(SMono.raiseError(BlobNotFoundException(blobId))))
+    }
+  }
+}
+
+class MessagePartBlobResolver @Inject()(val messageIdFactory: MessageId.Factory,
+                              val messageIdManager: MessageIdManager) extends BlobResolver {
+  private def asMessageAndPartId(blobId: BlobId): Try[(MessageId, PartId)] = {
+    blobId.value.value.split("_").toList match {
+      case List(messageIdString, partIdString) => for {
+        messageId <- Try(messageIdFactory.fromString(messageIdString))
+        partId <- PartId.parse(partIdString)
+      } yield {
+        (messageId, partId)
+      }
+      case _ => Failure(BlobNotFoundException(blobId))
+    }
+  }
+
+  override def resolve(blobId: BlobId, mailboxSession: MailboxSession): Option[SMono[Blob]] = {
```

Comment:
would it makes sense to make the return type 'SMono[Blob]' by returning SMono.empy if none ?
or None has a different meaning that a Some of an empty Mono in our usage ?

Reply:
It have. An empty mono means the blobId is not applicable to the resolver.

I did not find the single mono to be as efficient to carry out the intent.

Maybe a sealed class could help?